File: server-api/src/main/java/com/bytelegend/app/serverapi/data/Session.java
package com.bytelegend.app.serverapi.data;

import software.amazon.awssdk.enhanced.dynamodb.mapper.annotations.DynamoDbBean;
import software.amazon.awssdk.enhanced.dynamodb.mapper.annotations.DynamoDbPartitionKey;

import java.time.Duration;
import java.time.Instant;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

@DynamoDbBean
public class Session extends Mvcc {
    // A token expires after 14 days
    // If token is older than 7 days, renew the token
    public static final Duration DEFAULT_SESSION_AGE = Duration.ofDays(14);
    public static final Duration DEFAULT_SESSION_RENEW = Duration.ofDays(7);
    public static final Session NULL = new Session();
    private String id;
    private String playerId;
    private Instant createdAt;
    private Map<String, String> data = new HashMap<>();

    @DynamoDbPartitionKey
    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getPlayerId() {
        return playerId;
    }

    public void setPlayerId(String playerId) {
        this.playerId = playerId;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(Instant createdAt) {
        this.createdAt = createdAt;
    }

    public Map<String, String> getData() {
        return data;
    }

    public void setData(Map<String, String> data) {
        this.data = data;
    }

    public Long getTtl() {
        return getExpireTime().getEpochSecond();
    }

    public boolean isExpired() {
        return getExpireTime().compareTo(Instant.now()) < 0;
    }

    private Instant getExpireTime() {
        return createdAt.plus(DEFAULT_SESSION_AGE);
    }

    public boolean isRenewable() {
        return createdAt.plus(DEFAULT_SESSION_RENEW).compareTo(Instant.now()) < 0;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Session session = (Session) o;
        return Objects.equals(id, session.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }
}

File: server-api/src/main/kotlin/com/bytelegend/app/serverapi/data/Session.kt
package com.bytelegend.app.serverapi.data

import org.bson.codecs.pojo.annotations.BsonId
import org.bson.codecs.pojo.annotations.BsonIgnore
import java.time.Duration
import java.time.Instant
import java.util.HashMap

val DEFAULT_SESSION_AGE = Duration.ofDays(14)
val DEFAULT_SESSION_RENEW = Duration.ofDays(7)

class Session(
    @BsonId
    var id: String,
    var playerId: String,
) {
    var version = 0

    var createdAt: Instant = Instant.now()
    var data: MutableMap<String, String> = HashMap()

    val isRenewable: Boolean
        @BsonIgnore
        get() = createdAt.plus(DEFAULT_SESSION_RENEW) < Instant.now()

    val isExpired: Boolean
        @BsonIgnore
        get() = expireAt < Instant.now()

    val expireAt: Instant
        @BsonIgnore
        get() = createdAt.plus(DEFAULT_SESSION_AGE)

    override fun hashCode(): Int {
        return id.hashCode()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Session

        if (id != other.id) return false

        return true
    }
}
